### ISSUE TITLE
Clang build error for armv7a shared library

### ISSUE THREAD

**Issue by EngineerSmith** [curl/curl#4693]
### I did this
I attempted to build curl 7.54.1 (and 7.67.0) shared library (*.so) for android armv7a using the following command. As generated in config.log as I was using a [script](https://github.com/EngineerSmith/openssl-curl-android/blob/master/build-curl.sh).
> ./configure --host=armv7a-linux-androideabi --target=armv7a-linux-androideabi --prefix=/home/ubuntu/openssl-curl-android/curl/build/armeabi-v7a --with-ssl=/home/ubuntu/openssl-curl-android/curl/../openssl/build/armeabi-v7a --with-shared --disable-verbose --disable-manual --disable-crypto-auth --disable-unix-sockets --disable-ares --disable-rtsp --disable-ipv6 --disable-proxy --disable-versioned-symbols --enable-hidden-symbols --without-libidn --without-librtmp --without-zlib --disable-dict --disable-file --disable-ftp --disable-ftps --disable-gopher --disable-imap --disable-imaps --disable-pop3 --disable-pop3s --disable-smb --disable-smbs --disable-smtp --disable-smtps --disable-telnet --disable-tftp

The build was going well until it hit,
>clang: error: no such file or  directory: '../lib/.libs/libcurl.so' 

[Build log gist](https://gist.github.com/EngineerSmith/ebaaac13df2aac0e597672d444fc724e) ([Shortcut to error line](https://gist.github.com/EngineerSmith/ebaaac13df2aac0e597672d444fc724e#file-build-log-L990))

### I expected the following
libcurl.so built with my configuration for curl 7.54.1. As the other 3 configurations built the shared library (arm64-v8a, x86 android, and x86_64 android) in the same environment. 

### Versions
Git: `curl-7_54_1` and `curl-7_67_0` tags. I tried 7.67.0 to see if it patched, but the same issue appears.
NDK: `android-ndk-r20b` and `android-ndk-r19c`. Tried r19c with `curl-7_67_0` to see if it was an issue with the NDK, but the same issue appears.
NDK SDK 19 clang, but did try a few other SDK clang versions but resulted in the same issue. Tried using SDK 23 as INSTALL.md suggests for openssl too

### Operating system
> Distributor ID: Ubuntu
> Description:    Ubuntu 18.04.3 LTS
> Release:        18.04
> Codename:       bionic

**Comment by jay** [comment]
> --with-shared

From your script:

~~~shell
SHARED=disable

while getopts s opt
do
  case "${opt}" in
  s) SHARED=with;;
  [?]) echo "Usage: $0 [-s]\n\t-s | Compile shared curl library (*.so)"
       exit 1;;
  esac
done
~~~

That should be --enable-shared, since shared is the default it is by happenstance that you (should) get a shared library.

> clang: error: no such file or directory: '../lib/.libs/libcurl.so'

I don't see a cause but then again I'm not familiar with cross compiling for Android. Did you follow [Android build directions](https://github.com/curl/curl/blob/c7bc689/docs/INSTALL.md#android)? Are there any shared objects in lib/.libs ? Try `make V=1` for verbose output. You should see
`libtool: link:  gcc -shared  ......... libcurl.so.4.6.0`
`libtool: link: (cd ".libs" && rm -f "libcurl.so" && ln -s "libcurl.so.4.6.0" "libcurl.so")`

**Comment by EngineerSmith** [comment]
Thanks for the correct argument. I followed the Android build directions, settings up the environment variables and set up the correct structure for OpenSSL. [Verbose build log](https://gist.github.com/EngineerSmith/8ceb967948f1476c08bce5311029a52e#file-build_verbose-log) doesn't have `libtool: link:`, and I could not find [curl/lib](https://gist.github.com/EngineerSmith/8ceb967948f1476c08bce5311029a52e#file-ls-lib)/.lib as there was no .libs folder that I could see with `ls -a` within curl/lib

**Comment by EngineerSmith** [comment]
I just ran the build again and this time .libs was generated or I just missed it the first time. [Contents of lib/.lib](https://gist.github.com/EngineerSmith/8a1d52c33cfb7eab1c4f35e5916b4825)

**Comment by bagder** [comment]
I would recommend using the latest *release* tarball for this. That's the 7.67.0 with a shipped pre-built configure. 

>  Contents of lib/.libs

There's no `.so` version in there for some reason, only the static `.a`.

**Comment by EngineerSmith** [comment]
I did the build again with curl 6.76.0 release ([curl-7.67.0.tar.gz](https://github.com/curl/curl/releases/tag/curl-7_67_0)) and got these results. [Build log](https://gist.github.com/EngineerSmith/125b594f4ce3037dd26040ec97b4ef20#file-build_vernose_released_7-67-0-log) and [lib/.libs](https://gist.github.com/EngineerSmith/125b594f4ce3037dd26040ec97b4ef20#file-ls-a-lib)

**Comment by bagder** [comment]
My guess is then that libtool (or clang?) is acting up and for some reason doesn't create the shared library version for this build combo. The configure output clearly says it should and a typical build will. 

The command line to build the tool to use the shared libcurl library fails because there is none.

**Comment by EngineerSmith** [comment]
Update, clang packaged with NDK R21 (beta 2) also doesn't work. I'll be looking into libtool to see if there's anything I can do with that

**Comment by EngineerSmith** [comment]
I've tried changing libtool versions, but  I don't know much about it. I looked at OpenSSL to see if there was something wrong with it. I built different versions but still the same problem. I then tried building without OpenSSL and my other options, but I had no luck
> ./configure --host=armv7a-linux-androideabi --target=armv7a-linux-androideabi --prefix=/home/ubuntu/openssl-curl-android/curl/build/armeabi-v7a --enable-shared

I'm not sure what I could be doing wrong

**Comment by bagder** [comment]
Me neither, but configure, libtool and clang all work perfectly fine to build curl on several other platforms - like all the ones I ever use and try builds on.

**Comment by jay** [comment]
Try it with --disable-static see what happens. Also does this mean anything:
> checking for link... link -dump
> configure: WARNING: using cross tools not prefixed with host triplet

**Comment by EngineerSmith** [comment]
I added --disable-static and I get the same *.so is missing
[Build.log](https://gist.github.com/EngineerSmith/1478ebf7785891605c55cff7feea8503#file-build-log)
[.libs.txt](https://gist.github.com/EngineerSmith/1478ebf7785891605c55cff7feea8503#file-libs-txt)

**Comment by bagder** [comment]
I still don't think this is curl's (build) problem.

**Comment by jay** [comment]
Is it possible that your OpenSSL libs are preventing libtool from building a shared libcurl library? Like maybe PIC is missing or something. Check:

checking for OpenSSL headers version... unknown - 0x1010102fL
checking for OpenSSL library version... 1.1.0
checking for OpenSSL headers and library versions matching... no
configure: WARNING: OpenSSL headers and library versions do not match.
configure: WARNING: skipped the /dev/urandom detection when cross-compiling

-L/home/ubuntu/openssl-curl-android/curl/../openssl/build/armeabi-v7a/lib -lssl -lcrypto -L/home/ubuntu/openssl-curl-android/curl/../openssl/build/armeabi-v7a/lib

-I/home/ubuntu/openssl-curl-android/curl/../openssl/build/armeabi-v7a/include -I/home/ubuntu/openssl-curl-android/curl/../openssl/build/armeabi-v7a/include/openssl

Try configure with `--disable-shared --without-ssl` see what happens. Then I'm out of ideas you might try the [curl-library](https://cool.haxx.se/mailman/listinfo/curl-library) (libcurl) mailing list to ask if anyone has been successful building armv7a-linux-androideabi. The mailing list has a wider audience..

**Comment by EngineerSmith** [comment]
I fixed the OpenSSL version warning. I was trying to build OpenSSL for android19 while it needs at least version 23 for curl. `--without-ssl` causes the same issue. I'll try the mailing list to see if anyone else has had any success. I'll update here if I get a solution that works.

**Comment by EngineerSmith** [comment]
As I was writing my email, I saw the issue! I set up one of the environment variables incorrectly, retrospectively I should have checked. I tested it and I'm able to build libcurl.so for armv7a now! Thank you for trying to help me.

I set LD to "$TOOLCHAIN/bin/armv7a-linus-androideabi-ld" as that file does not exist it must have an effect on clang not allowing it to produce the file but gave no warning that it didn't exist. The correct path being "$TOOLCHAIN/bin/arm-linux-androideabi-ld"
